src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -461,27 +428,21 @@ void BaseCouplingScheme::checkCompletenessRequiredActions()
   }
 }
 
-void BaseCouplingScheme::setupDataMatrices(DataMap &data)
+void BaseCouplingScheme::setupDataMatrices()
 {
   PRECICE_TRACE();
-  PRECICE_DEBUG("Data size: {}", data.size());
+  PRECICE_DEBUG("Data size: {}", getAccelerationData().size());
   // Reserve storage for convergence measurement of send and receive data values
   for (ConvergenceMeasureContext &convMeasure : _convergenceMeasures) {
     PRECICE_ASSERT(convMeasure.couplingData != nullptr);
-    if (convMeasure.couplingData->oldValues.cols() < 1) {
-      utils::append(convMeasure.couplingData->oldValues,
-                    (Eigen::MatrixXd) Eigen::MatrixXd::Zero(convMeasure.couplingData->values().size(), 1));
-    }
+    convMeasure.couplingData->storeIteration();
   }
-  // Reserve storage for extrapolation of data values
-  if (_extrapolationOrder > 0) {
-    for (DataMap::value_type &pair : data) {
-      int cols = pair.second->oldValues.cols();
-      PRECICE_DEBUG("Add cols: {}, cols: {}", pair.first, cols);
-      PRECICE_ASSERT(cols <= 1, cols);
-      utils::append(pair.second->oldValues,
-                    (Eigen::MatrixXd) Eigen::MatrixXd::Zero(pair.second->values().size(), _extrapolationOrder + 1 - cols));
-    }
+  // Reserve storage for data values
+  for (DataMap::value_type &pair : getAccelerationData()) {
+    time::PtrWaveform       ptrWaveform(new time::Waveform(pair.second->values().size(), _extrapolationOrder + 2));
+    WaveformMap::value_type waveformPair = std::make_pair(pair.first, ptrWaveform);
+    _waveforms.insert(waveformPair);
+    pair.second->storeIteration(); // @ todo remove this duplicate of line above! But removing this line causes failure for mpirun --oversubscribe -np 4 ./testprecice -t PreciceTests/Serial/MultiCoupling
```

Comment:
TODO

Reply:
From https://github.com/precice/precice/pull/1004#discussion_r627345604:

> Is it correct to call `storeIteration()` for `convMeasure.couplingData` and `pair.second` (from `getAccelerationData()`?). This means that there are convergence measures without an acceleration and there are acceleration schemes without a convergence measure. We might end up calling `storeIteration()` twice on the same piece of data (which is no problem, but a bit confusing.). If this use-case is intended, we can also remove the todo above.

As soon as we figure this out I can solve or remove the todo.

Reply:
I like the change. Simplifies things.